sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,273 @@
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			"name":   name,
+			"target": target,
+		}).Error("Error establishing connection to gRPC server")
+		return nil, err
+	}
+
+	return &GRPCStreamingSpanSink{
+		grpcConn:   conn,
+		name:       name,
+		target:     target,
+		sinkClient: NewSpanSinkClient(conn),
+		commonTags: commonTags,
+		log:        log,
+	}, nil
+}
+
+// Start performs final preparations on the sink before it is
+// ready to begin ingesting spans.
+func (gs *GRPCStreamingSpanSink) Start(cl *trace.Client) error {
+	gs.traceClient = cl
+
+	var err error
+	gs.stream, err = gs.sinkClient.SendSpans(context.TODO())
+	if err != nil {
+		gs.log.WithFields(logrus.Fields{
+			"name":          gs.name,
+			"target":        gs.target,
+			"chanstate":     gs.grpcConn.GetState().String(),
+			logrus.ErrorKey: err,
+		}).Error("Failed to set up a stream over gRPC channel to sink target")
+		return err
+	}
+
+	go gs.maintainStream()
+	return nil
+}
+
+// maintainStream is intended to be run in a background goroutine. It is kicked
+// off by Start(), and is only kept as a standalone method for testing purposes.
+//
+// This method runs an endless loop that reacts to state changes in the
+// underlying channel by automatically attempting to re-establish the stream
+// connection when it moves back into the 'READY' state. See
+// https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.mdA
+// for details about gRPC's connectivity state machine.
+func (gs *GRPCStreamingSpanSink) maintainStream() {
+	for {
+		// This call will block on a channel receive until the gRPC connection
+		// becomes unhealthy. Then, spring into action!
+		state := gs.grpcConn.GetState()
+		switch state {
+		case connectivity.Idle:
+			// gRPC considers an open stream to be an active RPC, and channels
+			// normally only go idle if there are no active RPC within the
+			// channel's timeout window. It should be nigh-impossible for that
+			// to happen here, so we record entering the Idle state as an error.
+			gs.log.WithFields(logrus.Fields{
+				"name":          gs.name,
+				"target":        gs.target,
+				"chanstate":     gs.grpcConn.GetState().String(),
+				logrus.ErrorKey: fmt.Errorf("gRPC sink became idle; should be nearly impossible"),
+			}).Error("gRPC channel transitioned to idle")
+
+			// With the error sent, now fall through to recreating the stream
+			// as if the channel were ready, as stream creation should force
+			// the channel back out of idle. This isn't ideal - it'll end up
+			// recreating the stream again after re-reaching READY - but it's
+			// not a huge problem, as this case is essentially unreachable.
+			fallthrough
+		case connectivity.Ready:
+			gs.streamMut.Lock()
+			// CAS inside the mutex so that we can't ever get a double-send of the
+			// "first time channel degradation" log error.
+			atomic.CompareAndSwapUint32(&gs.bad, 1, 0)
+			stream, err := gs.sinkClient.SendSpans(context.TODO())
```

Comment:
what context should this use?

Reply:
good point - we can at least create an arg for this one, even though it's just pushing the `context.TODO()` up to a higher level :)